embassy-executor/src/arch/riscv32.rs

```diff
@@ -55,19 +55,11 @@ impl Executor {
             unsafe {
                 self.inner.poll();
                 // we do not care about race conditions between the load and store operations, interrupts
-                //will only set this value to true.
-                critical_section::with(|_| {
-                    // if there is work to do, loop back to polling
-                    // TODO can we relax this?
-                    if SIGNAL_WORK_THREAD_MODE.load(Ordering::SeqCst) {
-                        SIGNAL_WORK_THREAD_MODE.store(false, Ordering::SeqCst);
-                    }
-                    // if not, wait for interrupt
-                    else {
-                        core::arch::asm!("wfi");
-                    }
-                });
-                // if an interrupt occurred while waiting, it will be serviced here
+                // will only set this value to true.
+                // if there is work to do, loop back to polling
+                if !SIGNAL_WORK_THREAD_MODE.fetch_and(false, Ordering::SeqCst) {
+                    core::arch::asm!("wfi");
+                }
```

Comment:
the wfi in the critical section hangs on RISC-V, I think this code does the same, but I don't fully understand what it is for

Reply:
The critical section is to avoid a race where:
- Main thread checks SIGNAL_WORK_THREAD_MODE, sees it's `false`.
- Then an interrupt happens, sets `SIGNAL_WORK_THREAD_MODE` to `true`.
- Main thread goes to `wfi` anyway, because it had checked `SIGNAL_WORK_THREAD_MODE` before and seen `false`.
- `wfi` sleeps forever, the interrupt doesn't interrupt it because it has already happened in the past
- Everything hangs!

When this code was written, we thought that an interrupt becoming pending still causes `wfi` to exit, even if the interrupt doesn't immediately get handled because interrupts were disabled. Perhaps this is not the case? or perhaps only in some chips? iirc the riscv spec was somewhat tight-lipped regarding this.

cc @MabezDev @ivmarkov

Reply:
Reading riscv spec again makes me think you are right and wfi should just resume, even if interrupts are disabled (which my critical section implementation does).

However, I think my proposed change still makes sense; wfi would / should exit on pending interrupts, not on new interrupts alone. The critical section in the original code is not needed *I think* since we can atomically "fetch_and"?

Reply:
Your code is still vulnerable to the race if things execute in this order:

- main: does `fetch_and`, sees the flag is `false`
- irq: fires, sets the flag to `true`.
- main: does `wfi`, hangs.

This is why interrupts have to be disabled between checking the flag and `wfi`.

Reply:
Ah, I get it now, thanks!

I'll revert these changes, is the other stuff ok for merge? (then I'll rebase those on top of latest master)